external/storm-hive/src/main/java/org/apache/storm/hive/bolt/HiveBolt.java

```diff
@@ -104,16 +106,30 @@ public void execute(Tuple tuple) {
                 enableHeartBeatOnAllWriters();
             }
             writer.write(options.getMapper().mapRecord(tuple));
-            currentBatchSize++;
-            if(currentBatchSize >= options.getBatchSize()) {
+
+            tupleBatch.add(tuple);
+            if(tupleBatch.size() >= options.getBatchSize()) {
                 flushAllWriters();
-                currentBatchSize = 0;
+                LOG.info("acknowledging tuples after writers flushed ");
+                for(Tuple t : tupleBatch)
+                    collector.ack(t);
+                tupleBatch.clear();
             }
-            collector.ack(tuple);
+
         } catch(Exception e) {
             this.collector.reportError(e);
             collector.fail(tuple);
-            flushAndCloseWriters();
+            try {
+                flushAndCloseWriters();
+                LOG.info("acknowledging tuples after writers flushed and closed");
+                for (Tuple t : tupleBatch)
+                    collector.ack(t);
+                tupleBatch.clear();
+            } catch (Exception e1) {
+                //If flushAndClose fails assume tuples are lost, do not ack
+                LOG.warn("Error while flushing and closing writers, tuples will NOT be acknowledged");
```

Comment:
I am not quite sure but isn't it beneficial to calling fail instead of letting them timeout.